src/lib/digest_service/digest_data_extractor.py

```diff
@@ -113,12 +112,56 @@ def get_query(self, start_time: datetime, end_time: datetime) -> str:
 
 class GlueDataCatalogsDigestDataExtractor(BaseDigestDataExtractor):
     """
-    Class is responsible for preparing the query for extracting Glue Data Catalogs runs.
+    Class is responsible for preparing the query for extracting Glue Data Catalogs counts.
     """
 
     def get_query(self, start_time: datetime, end_time: datetime) -> str:
-        print("Calling a method which hasn't been implemented yet")
-        query = ""
+        query = f"""SELECT '{self.resource_type}' as resource_type
```

Comment:
The SQL might be frightening :)))
Maybe we consider doing something like (it's just pseudo-code):

WITH min_max_record (
  SELECT env, res_type, res_name, min(time), max(time) FROM tbl
    WHERE time BETWEEN '{start_time}' AND '{end_time}'
),
prep as (
SELECT 
       CASE WHEN tbl.time = min THEN tables_count ELSE 0 END as tables_prev,
       CASE WHEN tbl.time = max THEN tables_count ELSE 0 END as tables_current,       
   FROM tbl
   JOIN min_max_record on (env=, res_type=, res_name=, time in (min, max) 
)   
SELECT tables_current as table_count
          , tables_current - tables_prev as tables_diff
   FROM prep
   
   Just a guess. Maybe your option is better

Reply:
and, apart from that, I guess converting to CTE might make the query more readable comparing to nested queries

Reply:
thanks for your comments! at the moment not sure whether your solution would work, I think we still need to use LAG() window functions to get previous values

in case we convert subqueries to CTE, would this option be more readable?
WITH ranked_resources AS(
  SELECT 
  t.*
  , ROW_NUMBER() OVER (PARTITION BY resource_name ORDER BY time DESC) AS rn_desc
  , ROW_NUMBER() OVER (PARTITION BY resource_name ORDER BY time ASC) AS rn_asc
  FROM "timestream-salmon-metrics-events-storage-itary"."tstable-glue_catalogs-metrics" t
  -- WHERE time BETWEEN '{start_time}' AND '{end_time}'
),
-- filter to retain only the earliest and latest rows for each resource
filtered_resources AS (
    SELECT *
    FROM ranked_resources 
    WHERE rn_desc = 1 OR rn_asc = 1
),
-- get previous counts for tables, partitions, and indexes
counts AS (
    SELECT 
        monitored_environment,
        resource_name,
        tables_count,
        LAG(tables_count) OVER (PARTITION BY resource_name ORDER BY time) AS prev_tables_count,
        partitions_count,
        LAG(partitions_count) OVER (PARTITION BY resource_name ORDER BY time) AS prev_partitions_count,
        indexes_count,
        LAG(indexes_count) OVER (PARTITION BY resource_name ORDER BY time) AS prev_indexes_count,
        ROW_NUMBER() OVER (PARTITION BY resource_name ORDER BY time DESC) AS row_num
    FROM filtered_resources
)
-- final selection to calculate the deltas and include relevant fields
SELECT 
    'glue_catalogs' AS resource_type,
    monitored_environment,
    resource_name,
    tables_count,
    tables_count - prev_tables_count AS tables_added,
    partitions_count,
    partitions_count - prev_partitions_count AS partitions_added,
    indexes_count,
    indexes_count - prev_indexes_count AS indexes_added
FROM counts
WHERE row_num = 1